File: studio/Sources/StudioCore/GenerationSelection.swift
import Foundation

public enum GenerationProvider: String, CaseIterable, Identifiable {
  case local, drawThings
  public var id: String { rawValue }
  public var label: String { self == .local ? "WeeTodd (local)" : "Draw Things" }
}

extension Clip {
  public var reviewUsesSourceVideo: Bool {
    [.ltx23, .ltx25].contains(engine) && continuityMode == "motion"
  }
  public var reviewUsesContinuityFrame: Bool {
    engine != .drawThings && engine != .movie && continuityMode == "frame"
  }
  public var reviewAttachments: [Attachment] {
    attachments.filter {
      !($0.role == .first && reviewUsesContinuityFrame)
        && ($0.role != .lora || (engine != .drawThings && $0.isEnabled))
    }
  }
  public var reviewMediaCount: Int {
    return reviewAttachments.filter { $0.role != .lora }.count
      + (reviewUsesContinuityFrame || reviewUsesSourceVideo ? 1 : 0)
  }
  public var reviewLoRACount: Int {
    engine == .drawThings ? drawThings?.loras.filter(\.isEnabled).count ?? 0
      : reviewAttachments.filter { $0.role == .lora }.count
  }

  public mutating func selectAutomaticModelComponents() {
    let task = inferredTask
    profileID = "auto"
    if generationSelection == nil { generationSelection = GenerationSelection(task: task) }
  }

  public var generationProvider: GenerationProvider { engine == .drawThings ? .drawThings : .local }

  public mutating func selectGenerationProvider(_ provider: GenerationProvider) {
    guard provider != generationProvider else { return }
    if provider == .drawThings {
      selectGenerationEngine(.drawThings)
    } else {
      selectLocalModel([Engine.h3, .ltx23, .ltx25].contains(lastLocalEngine ?? .movie)
        ? lastLocalEngine! : .ltx25)
    }
  }

  public mutating func selectLocalModel(_ model: Engine) {
    guard [.h3, .ltx23, .ltx25].contains(model), model != engine else { return }
    let hasSavedSettings = savedNativeGenerations?[model.rawValue] != nil
    var media = self
    media.generationSelection = nil
    let task = media.inferredTask
    selectGenerationEngine(model)
    if !hasSavedSettings { generationSelection = GenerationSelection(task: task) }
  }

  public mutating func selectGenerationPreset(_ preset: GenerationPreset) {
    generationSelection = GenerationSelection(task: inferredTask, preset: preset)
  }

  public mutating func selectGenerationTask(_ task: String) {
    if generationSelection == nil { generationSelection = GenerationSelection() }
    generationSelection?.task = task
  }

  public mutating func selectGenerationEngine(_ target: Engine) {
    guard target != engine else { return }
    if [.h3, .ltx23, .ltx25].contains(engine) {
      lastLocalEngine = engine
      if savedNativeGenerations == nil { savedNativeGenerations = [:] }
      savedNativeGenerations?[engine.rawValue] = SavedNativeGenerationSettings(
        profileID: profileID, selection: generationSelection)
    }
    engine = target
    if let saved = savedNativeGenerations?[target.rawValue] {
      profileID = saved.profileID
      generationSelection = saved.selection
    } else {
      profileID = "auto"
      generationSelection = target == .drawThings || target == .movie ? nil : GenerationSelection()
    }
  }
}

/// Inactive native settings survive a backend round trip without affecting the current render.
public struct SavedNativeGenerationSettings: Codable, Equatable {
  public var profileID: String
  public var selection: GenerationSelection?
}

public enum GenerationPreset: String, Codable, CaseIterable, Identifiable {
  case balanced, speed, lowMemory, custom
  public var id: String { rawValue }
  public var label: String {
    switch self {
    case .balanced: return "Balanced"
    case .speed: return "Speed"
    case .lowMemory: return "Low memory"
    case .custom: return "Custom"
    }
  }
}

/// Explicit user intent. Nil on older clips preserves their exact recipe behavior.
public struct GenerationSelection: Codable, Equatable {
  public var task: String
  public var preset: GenerationPreset
  public var steps: Int?
  public var refinementSteps: Int?
  public var cfg: Double?
  public var shift: Double?
  public var memoryPolicy: String?
  public var projectionBackend: String?
  public init(task: String = "t2v", preset: GenerationPreset = .balanced) {
    self.task = task
    self.preset = preset
  }
  public var isModified: Bool {
    steps != nil || refinementSteps != nil || cfg != nil || shift != nil
      || memoryPolicy != nil || projectionBackend != nil
  }
  public mutating func resetOverrides() {
    steps = nil; refinementSteps = nil; cfg = nil; shift = nil
    memoryPolicy = nil; projectionBackend = nil
  }
  public static func taskLabel(_ task: String) -> String {
    switch task {
    case "t2v": return "Text to video"
    case "i2v": return "Image to video"
    case "fflf": return "First and last frames"
    case "ref2va": return "Reference video"
    case "a2v": return "Audio-driven video"
    case "control": return "Controlled video"
    case "extension": return "Video extension"
    default: return task
    }
  }
}

public struct GenerationControls: Codable, Equatable {
  public var evaluations: Int?
  public var refinementSteps: Int?
  public var cfg: Double?
  public var shift: Double?
  public var stepsEditable: Bool
  public var refinementStepsEditable: Bool
  public var cfgEditable: Bool
  public var shiftEditable: Bool
  public var stepsExplanation: String
  public var cfgExplanation: String
  public var shiftExplanation: String
}
public struct GenerationDescriptor: Codable, Equatable {
  public struct Preset: Codable, Equatable, Identifiable {
    public var id: String
    public var name: String
    public var description: String
  }
  public var supportedTasks: [String]
  public var controls: GenerationControls
  public var presets: [Preset]
}

public struct AccelerationSettings: Codable, Equatable {
  public var h3MemoryPolicy: String = "automatic"
  public var h3ProjectionBackend: String = "auto"
  public init() {}
}

extension GenerationSelection {
  /// Asset records can be relinked or reclassified without changing the clip's attachments.
  public static func assetFingerprint(for clip: Clip, assets: [MediaAsset]) -> String {
    let referenced = Set(clip.attachments.map(\.assetID))
    let records = assets.filter { referenced.contains($0.id) }
      .sorted { $0.id.uuidString < $1.id.uuidString }
    let encoder = JSONEncoder()
    encoder.outputFormatting = .sortedKeys
    return ((try? encoder.encode(records)) ?? Data()).base64EncodedString()
  }
}

extension AccelerationSettings {
  public static func memoryPolicyLabel(_ policy: String) -> String {
    switch policy {
    case "automatic": return "Automatic"
    case "recipe": return "Recipe default"
    case "paged": return "Paged · lower memory"
    case "pagedNormal": return "Paged · larger workspace"
    case "resident": return "Resident · experimental high RAM"
    default: return policy
    }
  }
}

File: src/wee_todd_mlx/generation_selection.py
"""Pure, lightweight generation selection shared by Studio and exported native jobs."""

from __future__ import annotations

import copy
import hashlib
import json
import math
from pathlib import Path


def infer_task(clip):
    if clip.get("extensionDirection"):
        return "extension"
    roles = {a["role"] for a in clip.get("attachments", [])}
    for role, task in (("control", "control"), ("audioDriver", "a2v"), ("reference", "ref2va")):
        if role in roles:
            return task
    return "fflf" if roles & {"first", "last", "keyframe"} else "t2v"


def supported_tasks(recipe):
    engine, config = recipe["engine"], recipe.get("config", {})
    components = recipe.get("components", {})
    if engine == "h3":
        tasks = {
            "t2va": ["t2v", "i2v", "fflf"],
            "fl2va": ["t2v", "i2v", "fflf"],
            "ref2va": ["ref2va", "a2v", "extension"],
        }.get(components.get("task"), [])
        if recipe.get("conditioning", {}).get("task") == "control":
            tasks = ["control"]
        if any(recipe.get(key) for key in ("attention", "fastvideo", "vdn")):
            tasks = [task for task in tasks if task == "t2v"]
        encoder = components.get("text_encoder")
        if encoder:
            manifest = Path(encoder).expanduser() / "paged_text_encoder_manifest.json"
            if manifest.is_file():
                if manifest.stat().st_size > 4 * 1024 * 1024:
                    raise ValueError("Text encoder provenance exceeds the 4 MiB inspection limit.")
                document = json.loads(manifest.read_text())
                if document.get("format") == "weetodd-h3-qwen-paged-v1":
                    tasks = [task for task in tasks if task == "t2v"]
        return tasks
    if engine == "ltx23":
        ic = components.get("ic_loras", [])
        if ic:
            return ["ref2va" if ic[0].get("family") == "ingredients_reference_sheet" else "control"]
        return (
            ["t2v", "i2v", "fflf", "a2v"]
            if config.get("pipeline_mode", "two_stage") == "two_stage"
            else ["t2v", "extension"]
            if config.get("pipeline_mode") in {"one_stage", "distilled"}
            else ["t2v"]
        )
    if engine == "ltx25":
        task = recipe.get("conditioning", {}).get("task", "t2v")
        if (
            components.get("ic_loras")
            or config.get("ic_lora_single_stage")
            or task in {"control", "ref2va"}
        ):
            return [task]
        return ["t2v", "i2v", "fflf", "a2v"] + (
            ["extension"] if config.get("pipeline_mode", "distilled") == "distilled" else []
        )
    return []


def h3_checkpoint_sampling(recipe):
    """Inspect bounded provenance JSON, never tensor payloads or checkpoint filenames."""
    if recipe.get("engine") != "h3":
        return {}
    component = recipe.get("components", {}).get("transformer")
    if not component:
        return {}
    root = Path(component).expanduser()
    result = {}
    for name in ("paged_manifest.json", "model_identity.json", "conversion_provenance.json"):
        source = root / name
        if not source.is_file():
            continue
        if source.stat().st_size > 4 * 1024 * 1024:
            raise ValueError("Checkpoint provenance exceeds the 4 MiB inspection limit.")
        document = json.loads(source.read_text())
        if not isinstance(document, dict):
            raise ValueError("Checkpoint provenance must be a JSON object.")
        records = [document]
        if isinstance(document.get("metadata"), dict):
            records.append(document["metadata"])
        for record in records:
            declared_source = str(record.get("source", "")).lower()
            if any(marker in declared_source for marker in ("fasth3", "vdn-h3", "turbo")):
                result["special"] = True
            sampling = record.get("sampling")
            if isinstance(sampling, dict) and sampling:
                result["special"] = True
                for key in ("schedule_points", "transformer_evaluations"):
                    value = sampling.get(key)
                    if value is not None:
                        if type(value) is not int or value < 1:
                            raise ValueError(
                                "Checkpoint sampling counts must be positive integers."
                            )
                        if key in result and result[key] != value:
                            raise ValueError(
                                "Checkpoint provenance has conflicting sampling counts."
                            )
                        result[key] = value
    return result


def special_h3_sampling(recipe):
    if recipe.get("engine") != "h3":
        return False
    if h3_checkpoint_sampling(recipe).get("special"):
        return True
    if any(recipe.get(key) for key in ("attention", "fastvideo", "vdn")):
        return True
    for adapter in recipe.get("loras", {}).get("adapters", []):
        if adapter.get("profile") == "standard":
            continue
        if adapter.get("profile") == "turbo":
            return True
        try:
            from wee_todd_nodes.lora import H3LoRASpec

            if H3LoRASpec(**adapter).resolved_profile != "standard":
                return True
        except (OSError, ValueError, TypeError, KeyError):
            # Unknown adapters must not make the generic schedule editable.
            return True
    return False


def generation_descriptor(recipe):
    engine, config = recipe["engine"], recipe.get("config", {})
    h3 = engine == "h3"
    mode = config.get("pipeline_mode", "distilled" if engine == "ltx25" else "two_stage")
    ordinary = (
        h3
        and not special_h3_sampling(recipe)
        and config.get("sampling_method", "euler") == "euler"
        and not any(config.get(key) for key in ("sigmas", "custom_sigmas", "turbo"))
    )
    steps = config.get("steps", 16) - 1 if ordinary else None
    checkpoint_sampling = h3_checkpoint_sampling(recipe)
    turbo_schedule = h3 and any(
        adapter.get("profile") == "turbo"
        for adapter in recipe.get("loras", {}).get("adapters", [])
    ) and config.get("sampling_method", "euler") == "euler"
    if turbo_schedule:
        steps = config.get("steps", 16) - 1
    if h3 and checkpoint_sampling.get("transformer_evaluations"):
        steps = checkpoint_sampling["transformer_evaluations"]
    refinement = None
    if not h3:
        steps = config.get("stage1_steps", 8 if engine == "ltx25" else 30)
        refinement = (
            config.get("stage2_steps", 3)
            if mode not in {"one_stage", "distilled_single_stage"}
            else None
        )
        if engine == "ltx25" and config.get("stage1_sampler") == "euler_ancestral_cfg_pp":
            from ltx25_mlx.runtime import LTX25_CFG_PP_SCHEDULES

            steps += len(LTX25_CFG_PP_SCHEDULES[config.get("cfg_pp_schedule", "full")])
    ic = recipe.get("components", {}).get("ic_loras", []) if engine == "ltx23" else []
    if engine == "ltx23" and (mode == "distilled" or ic):
        steps = min(steps, 8)
        refinement = min(refinement, 3) if refinement is not None else None
        topology = config.get("ic_lora_topology", "auto")
        if ic and topology == "control_refine":
            refinement = min(config.get("stage2_steps", 3), 8)
        if topology in {"single_stage", "upsample_only"} or (
            topology == "auto" and ic and ic[0].get("family") == "motion_track"
        ):
            refinement = None
    single = engine == "ltx23" and mode == "distilled_single_stage"
    cfg_editable = not h3 and mode not in {"distilled", "distilled_single_stage"} and not ic
    controls = {
        "evaluations": steps,
        "refinementSteps": refinement,
        "cfg": None
        if h3 or ic
        else config.get("video_cfg_scale", 1.0)
        if engine == "ltx25"
        else config.get("cfg_scale", 3.0),
        "shift": config.get("shift", 5) if single else None,
        "stepsEditable": ordinary or (engine == "ltx23" and mode != "distilled" and not ic),
        "refinementStepsEditable": engine == "ltx23"
        and refinement is not None
        and mode != "distilled"
        and not ic,
        "cfgEditable": cfg_editable,
        "shiftEditable": single,
        "stepsExplanation": (
            f"Turbo sampling · {steps} evaluations. Disable Turbo to restore standard Steps."
        )
        if turbo_schedule
        else "Actual Euler evaluations; the recipe stores one extra schedule point."
        if ordinary
        else "Actual full-resolution evaluations; the tested setting is 8."
        if single
        else "Stage-one schedule steps. Fixed schedules cannot be overridden."
        if not h3
        else "Fixed checkpoint schedule; preserve its declared evaluation count."
        if checkpoint_sampling.get("transformer_evaluations")
        else "Custom schedule: evaluation count is not inferred.",
        "cfgExplanation": "Native H3 uses distilled guidance; CFG overrides are unsupported."
        if h3
        else "Video CFG; audio CFG retains the recipe value."
        if engine == "ltx25" and cfg_editable
        else "Distilled guidance is fixed."
        if not cfg_editable
        else "Guidance scale for the selected native pipeline.",
        "shiftExplanation": "Linear trailing schedule Shift; the tested setting is 5."
        if single
        else "This native adapter does not expose a Shift override.",
    }
    return {
        "supportedTasks": supported_tasks(recipe),
        "controls": controls,
        "presets": [
            {"id": "custom", "name": "Custom", "description": "Preserve imported recipe settings."},
            {
                "id": "balanced",
                "name": "Balanced",
                "description": "Preserve validated recipe sampling settings.",
            },
            {
                "id": "speed",
                "name": "Speed",
                "description": "Keep recipe sampling pending faster-settings qualification.",
            },
            {
                "id": "lowMemory",
                "name": "Low memory",
                "description": "Staged loading; hardware qualification remains separate.",
            },
        ],
    }


def fingerprint(recipe):
    return hashlib.sha256(
        json.dumps(recipe, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()
    ).hexdigest()


def resolve_generation_selection(selection, clip, profiles, capabilities, *, validate_inputs=True):
    """Resolve recipe content, never filename hints; leave caller-owned objects untouched."""
    selection = dict(selection or {})
    unknown = set(selection) - {
        "task",
        "preset",
        "steps",
        "refinementSteps",
        "cfg",
        "shift",
        "memoryPolicy",
        "projectionBackend",
    }
    if unknown:
        raise ValueError(f"Unsupported generation controls: {sorted(unknown)}")
    explicit = bool(selection)
    acceleration = None
    if explicit and clip["engine"] == "h3" and selection.get("preset", "custom") != "custom":
        from wee_todd_mlx.acceleration import resolve_h3_acceleration

        acceleration = resolve_h3_acceleration(
            capabilities.get("acceleration"), capabilities.get("hardware")
        )
        selection.setdefault("projectionBackend", acceleration["projectionBackend"])
        if selection.get("preset") != "lowMemory":
            selection.setdefault("memoryPolicy", acceleration["memoryPolicy"])
    task = selection.get("task") or infer_task(clip)
    required = {
        "i2v": ["first"],
        "fflf": ["first", "last"],
        "ref2va": ["reference"],
        "a2v": ["audioDriver"],
        "control": ["control"],
        "extension": ["source"],
    }.get(task, [])
    readiness_errors = []
    if explicit:
        roles = {a["role"] for a in clip.get("attachments", [])} - {"lora"}
        labels = {
            "i2v": "Image to video",
            "fflf": "First and last frames",
            "ref2va": "Reference to video",
            "a2v": "Audio to video",
            "control": "Control to video",
            "extension": "Video extension",
        }
        input_labels = {
            "first": "First frame image",
            "last": "Last frame image",
            "reference": "Reference media",
            "audioDriver": "Audio driver",
            "control": "Control guide",
            "source": "Source movie",
        }
        missing = []
        for role in required:
            present = (
                bool(clip.get("extensionSource") or clip.get("sourcePath"))
                if role == "source"
                else role in roles
            )
            if not present:
                missing.append(input_labels[role])
        if missing:
            readiness_errors.append(f"{labels.get(task, task)} requires {' and '.join(missing)}.")
        allowed = {
            "t2v": set(),
            "i2v": {"first"},
            "fflf": {"first", "last", "keyframe"},
            "ref2va": {"reference"},
            "a2v": {"audioDriver", "first", "last", "keyframe"},
            "control": {"control"},
            "extension": set(),
        }.get(task, set())
        conflicts = roles - allowed
        if conflicts:
            readiness_errors.append(
                f"{task} conflicts with attached {', '.join(sorted(conflicts))}; "
                "attachments are preserved. Remove them or select their task."
            )
    if validate_inputs and readiness_errors:
        raise ValueError(readiness_errors[0])
    candidates = []
    for profile in profiles:
        recipe = profile.get("recipe")
        if recipe is None:
            recipe = json.loads(Path(profile["id"]).read_text())
        if recipe.get("engine") == clip["engine"]:
            candidates.append((profile, recipe))
    selected = clip.get("profileID") or "auto"
    if selected != "auto":
        candidates = [(p, r) for p, r in candidates if p["id"] == selected]
        if not candidates:
            raise ValueError("The selected model recipe is missing. Reimport or select Automatic.")
    compatible = [(p, r) for p, r in candidates if task in supported_tasks(r)]
    if not explicit and clip["engine"] == "h3":
        native_task = {
            "t2v": "t2va",
            "fflf": "fl2va",
            "ref2va": "ref2va",
            "a2v": "ref2va",
            "extension": "ref2va",
            "control": "t2va",
        }.get(task)
        compatible = [
            (p, r) for p, r in compatible if r.get("components", {}).get("task") == native_task
        ]
    if clip["engine"] == "ltx25" and task == "control":
        from wee_todd_mlx.model_setup import ltx25_recipe_control_families
        from wee_todd_mlx.task_conditioning import CONTROL_FAMILIES

        required_families = {
            CONTROL_FAMILIES.get(a.get("controlType", "canny_edges"))
            for a in clip.get("attachments", [])
            if a["role"] == "control"
        }
        checked = []
        for candidate, content in compatible:
            try:
                families = ltx25_recipe_control_families(candidate["id"])
            except (OSError, ValueError, KeyError, TypeError):
                continue
            if None not in required_families and required_families <= families:
                checked.append((candidate, content))
        compatible = checked
    if not compatible:
        raise ValueError(
            f"{clip['engine']} {task} conflicts with the selected recipe. "
            "Import a compatible recipe or select Automatic. No inputs were discarded."
        )
    if explicit and selected == "auto":
        def preference(pair):
            content = pair[1]
            native_task = "fflf" if task == "i2v" else task
            familiar = selection.get("preset", "custom") != "custom"
            return (
                special_h3_sampling(content),
                familiar and content.get("conditioning", {}).get("task", "t2v") != native_task,
                familiar and clip["engine"] in {"ltx23", "ltx25"}
                and content.get("config", {}).get("pipeline_mode")
                not in {"distilled", "distilled_single_stage"},
            )

        compatible.sort(key=preference)
        if selection.get("preset", "custom") != "custom" and special_h3_sampling(compatible[0][1]):
            raise ValueError("Import a standard native recipe for this preset, or select Custom.")
    profile, original = compatible[0]
    recipe = copy.deepcopy(original)
    if explicit and clip["engine"] == "h3" and task in {"t2v", "i2v", "fflf"}:
        recipe["components"]["task"] = "t2va" if task == "t2v" else "fl2va"
    if explicit:
        contract = recipe.setdefault("conditioning", {})
        resolved_task = "fflf" if task == "i2v" else task
        if contract.get("task") != resolved_task:
            contract.pop("audio_policy", None)
        contract["task"] = resolved_task
    turbo_warnings = []
    if clip["engine"] == "h3" and capabilities.get("attached_loras"):
        from wee_todd_mlx.studio_h3_turbo import resolve_studio_h3_turbo

        recipe, turbo_warnings = resolve_studio_h3_turbo(
            recipe, capabilities["attached_loras"], task
        )
    config = recipe.setdefault("config", {})
    checkpoint_sampling = h3_checkpoint_sampling(recipe)
    fixed_points = checkpoint_sampling.get("schedule_points")
    if fixed_points is not None and config.get("steps", 16) != fixed_points:
        raise ValueError(f"This checkpoint requires a fixed schedule of {fixed_points} points.")
    preset = selection.get("preset", "custom")
    if preset not in {"custom", "balanced", "speed", "lowMemory"}:
        raise ValueError(f"Unknown generation preset: {preset}")

    def describe_resolved():
        inspected = copy.deepcopy(recipe)
        if clip["engine"] == "h3" and capabilities.get("attached_loras"):
            inspected.setdefault("loras", {}).setdefault("adapters", []).extend(
                capabilities["attached_loras"]
            )
        return generation_descriptor(inspected)

    descriptor = describe_resolved()
    controls = descriptor["controls"]
    for key in ("steps", "refinementSteps", "cfg", "shift"):
        value = selection.get(key)
        if value is None:
            continue
        if key == "steps" and turbo_warnings:
            # Preserve the saved standard override so disabling Turbo restores it.
            # The effective Turbo count is validated and reported by its resolver.
            continue
        if not controls[key + "Editable"]:
            raise ValueError(
                f"{key} override is unsupported. "
                + controls.get(key + "Explanation", "The selected native schedule is fixed.")
            )
        if (
            type(value) not in {int, float}
            or not math.isfinite(value)
            or value < (0 if key == "cfg" else 1)
        ):
            raise ValueError(f"{key} must be a finite valid positive value.")
        if key in {"steps", "refinementSteps"} and (type(value) is not int or value > 1000):
            raise ValueError(f"{key} must be an integer between 1 and 1000.")
        target = {
            "steps": "steps" if clip["engine"] == "h3" else "stage1_steps",
            "refinementSteps": "stage2_steps",
            "cfg": "video_cfg_scale" if clip["engine"] == "ltx25" else "cfg_scale",
            "shift": "shift",
        }[key]
        config[target] = value + 1 if key == "steps" and clip["engine"] == "h3" else value
    policy = selection.get("memoryPolicy", "paged" if preset == "lowMemory" else "recipe")
    if policy not in {"recipe", "paged", "pagedNormal", "resident"}:
        raise ValueError("Unknown memoryPolicy; use recipe, paged, pagedNormal, or resident.")
    if policy == "pagedNormal" and clip["engine"] != "h3":
        raise ValueError("pagedNormal is supported for native H3 only.")
    if policy != "recipe":
        if clip["engine"] == "h3":
            if policy == "resident" and config.get("paging_cache_gb", 0) > 0:
                raise ValueError("Resident sampling requires paging cache budget zero.")
            config["memory_mode"] = (
                "normal" if policy in {"resident", "pagedNormal"} else "low_memory_bf16"
            )
            recipe["block_residency"] = "resident" if policy == "resident" else "checkpoint_default"
        else:
            config.update(low_memory=True, low_ram_streaming=policy == "paged")
    backend = selection.get("projectionBackend")
    if backend is not None:
        if clip["engine"] != "h3" or backend not in {"mlx", "auto"}:
            raise ValueError("projectionBackend supports mlx or auto for native H3 only.")
        config["projection_backend"] = backend
    descriptor = describe_resolved()
    if acceleration:
        from wee_todd_mlx.acceleration import effective_h3_acceleration_report

        descriptor["acceleration"] = effective_h3_acceleration_report(
            acceleration, policy, config.get("projection_backend", "auto")
        )
    warnings = (
        ["Speed keeps the recipe sampling settings pending measured qualification."]
        if preset == "speed" and not turbo_warnings
        else []
    )
    warnings.extend(turbo_warnings)
    return {
        "recipe": recipe,
        "resolved_controls": descriptor["controls"],
        "generation": descriptor,
        "required_inputs": required,
        "readiness_errors": readiness_errors,
        "warnings": warnings,
        "fingerprint": fingerprint(recipe),
        "profileID": profile["id"],
        "task": task,
    }
